icu4c/source/test/intltest/unitstest.cpp

```diff
@@ -242,10 +250,274 @@ void UnitsTest::testConversions() {
+        if (U_SUCCESS(errorCode)) {
+            _amounts[_compoundCount] = dqOutputD.toDouble();
+            return true;
+        } else if (errorCode == U_DECIMAL_NUMBER_SYNTAX_ERROR) {
+            errorCode = U_ZERO_ERROR;
+        } else {
+            return false;
+        }
+
+        _measureUnits[_compoundCount] = MeasureUnit::forIdentifier(output, errorCode);
+        if (U_SUCCESS(errorCode)) {
+            _compoundCount++;
+            _skippedFields = 0;
+            return true;
+        }
+        _skippedFields++;
+        if (_skippedFields < 2) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Produces an output string for debug purposes.
+     * @param dbgOutput the buffer to which to write the output string.
+     * @param dbgOutputSize the size of the buffer. 
+     * @returns a null-terminated string containing each amount and unit.
+     */
+    void toDebugStringN(char *dbgOutput, int dbgOutputSize) {
+        char *next = dbgOutput;
+        char *end = dbgOutput + dbgOutputSize;
+        for (int i = 0; i < _compoundCount; i++) {
+            int printlen =
+                snprintf(next, dbgOutputSize, "%f %s ", _amounts[i], _measureUnits[i].getIdentifier());
+            next += printlen;
+            dbgOutputSize -= printlen;
+            if (dbgOutputSize <= 0) { break; }
+        }
+        // Terminate string with a null.
+        if (next >= end) {
+            dbgOutput[dbgOutputSize - 1] = 0;
+        } else {
+            *next = 0;
+        }
+    }
+};
+
+/**
+ * WIP(hugovdm): deals with a single data-driven unit test for unit preferences.
+ * This is a UParseLineFn as required by u_parseDelimitedFile.
+ */
+void unitPreferencesTestDataLineFn(void *context, char *fields[][2], int32_t fieldCount,
+                                   UErrorCode *pErrorCode) {
+    if (U_FAILURE(*pErrorCode)) return;
+    UnitsTest *intltest = (UnitsTest *)context;
+    IcuTestErrorCode status(*(UnitsTest *)context, "unitPreferencesTestDatalineFn");
+
+    if (!intltest->assertTrue(u"unitPreferencesTestDataLineFn expects 9 fields for simple and 11 "
+                              u"fields for compound. Other field counts not yet supported. ",
+                              fieldCount == 9 || fieldCount == 11)) {
+        return;
+    }
+
+    StringPiece quantity = trimField(fields[0]);
+    StringPiece usage = trimField(fields[1]);
+    StringPiece region = trimField(fields[2]);
+    StringPiece inputR = trimField(fields[3]);
+    StringPiece inputD = trimField(fields[4]);
+    StringPiece inputUnit = trimField(fields[5]);
+    ExpectedOutput output;
+    for (int i = 6; i < fieldCount; i++) {
+        output.parseOutputField(trimField(fields[i]));
+    }
+
+    DecimalQuantity dqInputD;
+    dqInputD.setToDecNumber(inputD, status);
+    if (status.errIfFailureAndReset("parsing decimal quantity: \"%.*s\"", inputD.length(),
```

Comment:
".*" is a precision field with a "dynamic value" which is passed as a parameter - thus two parameters: length then data, such that it's okay for data to not be a null-terminated string. An alternative would be to convert it to a null-terminated string - I don't think that's cleaner though.

Reply:
(Precision field on Wikipedia: https://en.wikipedia.org/wiki/Printf_format_string#Precision_field - because I've not decided what's my favourite C++ Docs website yet.)